File: demo/python/lunar_eclipse.py
#!/usr/bin/env python3
#
#    lunar_eclipse.py  -  by Don Cross - 2020-05-16
#
#    Example Python program for Astronomy Engine:
#    https://github.com/cosinekitty/astronomy
#
#    Searches for the next 10 partial/total lunar eclipses after
#    the current date, or a date specified on the command line.
#
#    To execute, run the command:
#    python3 lunar_eclipse.py [date]
#
import sys
from astronomy import Time, SearchLunarEclipse, NextLunarEclipse, EclipseKind


def PrintEclipse(e):
    # Calculate beginning/ending of different phases
    # of an eclipse by subtracting/adding the peak time
    # with the number of minutes indicated by the "semi-duration"
    # fields sd_partial and sd_total.
    MINUTES_PER_DAY = 24 * 60
    p1 = e.peak.AddDays(-e.sd_partial / MINUTES_PER_DAY)
    print('{}  Partial eclipse begins.'.format(p1))
    if e.sd_total > 0.0:
        t1 = e.peak.AddDays(-e.sd_total / MINUTES_PER_DAY)
        print('{}  Total eclipse begins.'.format(t1))
    print('{}  Peak of {} eclipse.'.format(e.peak, e.kind.name.lower()))
    if e.sd_total > 0.0:
        t2 = e.peak.AddDays(+e.sd_total / MINUTES_PER_DAY)
        print('{}  Total eclipse ends.'.format(t2))
    p2 = e.peak.AddDays(+e.sd_partial / MINUTES_PER_DAY)
    print('{}  Partial eclipse ends.'.format(p2))
    print()


def main(args):
    if len(args) == 1:
        time = Time.Now()
    elif len(args) == 2:
        time = Time.Parse(args[1])
    else:
        print('USAGE: {} [yyyy-mm-ddThh:mm:ssZ]'.format(args[0]))
        return 1

    count = 0
    e = SearchLunarEclipse(time)
    while count < 10:
        if e.kind != EclipseKind.Penumbral:
            count += 1
            PrintEclipse(e)
        e = NextLunarEclipse(e.peak)


if __name__ == '__main__':
    sys.exit(main(sys.argv))

File: demo/nodejs/lunar_eclipse.js
/*
    lunar_eclipse.js  -  by Don Cross - 2020-05-17

    Example Node.js program for Astronomy Engine:
    https://github.com/cosinekitty/astronomy

    Searches for the next 10 partial/total lunar eclipses after
    the current date, or a date specified on the command line.

    To execute, run the command:
    node lunar_eclipse [date]
*/

const Astronomy = require('../../source/js/astronomy.js');

function Pad(s, w) {
    s = s.toFixed(0);
    while (s.length < w) {
        s = '0' + s;
    }
    return s;
}

function FormatDate(t) {
    const date = t.date;
    var year = Pad(date.getUTCFullYear(), 4);
    var month = Pad(1 + date.getUTCMonth(), 2);
    var day = Pad(date.getUTCDate(), 2);
    var hour = Pad(date.getUTCHours(), 2);
    var minute = Pad(date.getUTCMinutes(), 2);
    var svalue = date.getUTCSeconds() + (date.getUTCMilliseconds() / 1000);
    var second = Pad(Math.round(svalue), 2);
    return `${year}-${month}-${day} ${hour}:${minute}:${second} UTC`;
}

function PrintEclipse(e) {
    // Calculate beginning/ending of different phases
    // of an eclipse by subtracting/adding the peak time
    // with the number of minutes indicated by the "semi-duration"
    // fields sd_partial and sd_total.
    const MINUTES_PER_DAY = 24 * 60;

    const p1 = e.peak.AddDays(-e.sd_partial / MINUTES_PER_DAY);
    console.log(`${FormatDate(p1)} - Partial eclipse begins.`);

    if (e.sd_total > 0) {
        const t1 = e.peak.AddDays(-e.sd_total / MINUTES_PER_DAY);
        console.log(`${FormatDate(t1)} - Total eclipse begins.`);
    }

    console.log(`${FormatDate(e.peak)} - Peak of ${e.kind} eclipse.`);

    if (e.sd_total > 0) {
        const t2 = e.peak.AddDays(+e.sd_total / MINUTES_PER_DAY);
        console.log(`${FormatDate(t2)} - Total eclipse ends.`);
    }

    const p2 = e.peak.AddDays(+e.sd_partial / MINUTES_PER_DAY);
    console.log(`${FormatDate(p2)} - Partial eclipse ends.`);
    console.log('');
}

function ParseDate(text) {
    const d = new Date(text);
    if (!Number.isFinite(d.getTime())) {
        console.error(`ERROR: Not a valid date: "${text}"`);
        process.exit(1);
    }
    return d;
}

function Demo() {
    const date = (process.argv.length === 3) ? ParseDate(process.argv[2]) : new Date();
    let count = 0;
    let eclipse = Astronomy.SearchLunarEclipse(date);
    for(;;) {
        if (eclipse.kind !== 'penumbral') {
            PrintEclipse(eclipse);
            if (++count === 10) {
                break;
            }
        }
        eclipse = Astronomy.NextLunarEclipse(eclipse.peak);
    }
}

Demo();
process.exit(0);
